src/sys/ptrace/linux.rs

```diff
@@ -310,6 +310,9 @@ fn ptrace_peek(
 /// `ptrace(PTRACE_GETREGSET, pid, NT_PRSTATUS, ...)` is used instead to achieve the same effect
 /// on aarch64 and riscv64.
 ///
+/// Currently, in x86_64 platform, if the tracer is 64bit and tracee is 32bit, this function
+/// will return an error.
```

Comment:
Does it?  Did you forget to add a size check below?

Reply:
Oh yes, I forgot this check. But as stated in the man page, PTRACE_GETREGS does not return the size of bytes written, so it seems that there is no easy way to check this?

Reply:
Do you know how C programs use PTRACE_GETREGS for 32-bit targets, then?

Reply:
As far as I know, we should check the target tracee bitness in advance, instead of checking after the syscall returns. Is it better to mark this method unsafe and provide a safe wrapper which does the bitness check?

Reply:
If it cannot guarantee that the return value will be correctly constructed, then yes it must be unsafe.  Either that, or it could return a union (which is unsafe to read from).

Reply:
Updated. I prefer to mark it as unsafe, since it is hard to import the 32bit user_regs_struct in `libc` crate.

Reply:
> As far as I know, we should check the target tracee bitness in advance, instead of checking after the syscall returns. 

From this [post](https://unix.stackexchange.com/questions/106234/determine-if-a-specific-process-is-32-or-64-bit), it is not trivial to check the bitness, so I would also prefer to mark this function unsafe as well if we don't have a better option.

 > since it is hard to import the 32bit user_regs_struct in libc crate

Would you like to elaborate on this a bit?

Reply:
If we return a union, then this union should involve both 64bit and 32bit version of user_regs_struct, which is provided by libc crate instead of the nix itself. However, the struct definition is bound to the compile target of libc crate, we cannot import a 32bit user_regs_struct in libc if we compiles to x86_64

Reply:
> However, the struct definition is bound to the compile target of libc crate, we cannot import a 32bit user_regs_struct in libc if we compiles to x86_64

Right, the actual type is decided at compile time, so we cannot use the 32-bit version unless we define it in Nix, which is not something Nix should do.

```sh
$ pwd
libc/src/unix/linux_like/linux/gnu

$ rg "struct user_regs_struct"
b64/x86_64/mod.rs
178:    pub struct user_regs_struct {

b32/x86/mod.rs
74:    pub struct user_regs_struct {

b64/loongarch64/mod.rs
193:    pub struct user_regs_struct {

b64/riscv64/mod.rs
196:    pub struct user_regs_struct {

b32/riscv32/mod.rs
200:    pub struct user_regs_struct {

b64/aarch64/mod.rs
146:    pub struct user_regs_struct {
```

------

> As far as I know, we should check the target tracee bitness in advance, instead of checking after the syscall returns.

After figuring out the bitness, how do you use a 32-bit version structure in C if you know the tracee is a 32-bit process? Since in C, these types are also chosen at compile time with `#ifdef` macros.

Reply:
In GDB, it is done without the user_regs_struct. The regs are fetched [here](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gdb/amd64-linux-nat.c;h=aa9295d57230d4f559f6b5c346658b786c0b0688;hb=HEAD#l222), and is used [here](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gdb/amd64-nat.c;h=5c5c368c93d64671a86bf1778cc8cef497a7368e;hb=HEAD#l86)

In short, it provides a 64bit version buffer to receive the regs, and when storing it to internal structure, the tracee arch is checked and collect corresponding register values.